### ISSUE TITLE
the script deletes all files within . if not able to read the variables properly due to lack of permissions

### ISSUE THREAD

**Issue by theruck242** [virtualox/Clear-CCMCache.ps1#2]
when you run the script by mistake under a non privileged powershell, the script deletes all the files in  . (the folder where you run the script from) 
![image](https://github.com/user-attachments/assets/cc40128b-cc4d-4635-8517-d4627608d3b0)

possibly the same result will be with privileged shell just with other reason not being able to read the $CacheFolderDisk with an actual path

**Comment by theruck242** [comment]
you can see what it does when you edit this line
`    # Remove orphaned folders from disk
    $CacheFoldersDisk | ForEach-Object { if ($_ -notin $CacheFoldersWMI.Location) { write-host $_ Remove-Item -Path $_ -Recurse -Force -ErrorAction SilentlyContinue } }`
which results in
C:\scripts\New folder\ClearCCMCache.ps1 Remove-Item -Path C:\scripts\New folder\ClearCCMCache.ps1 -Recurse -Force

**Comment by theruck242** [comment]
it looks that the 
`$CacheFoldersDisk = (Get-ChildItem $CachePath).FullName`
returns . when it fails to read the WMI object here
`$CachePath = ([wmi]"ROOT\ccm\SoftMgmtAgent:CacheConfig.ConfigKey='Cache'").Location`
so this deserves some error handling to prevent unvwanted deletions

**Comment by theruck242** [comment]
i edited it so that it
- detects powershell version as in 7.x the get-wmiobject does not work
- fixes the positionalbinding parameter
- fixes the deletion in case of non retrieval of the $Cachepath variable

```
<#
.SYNOPSIS
Clears old and unused content from the CCMCache folder.

.DESCRIPTION
This script detects and removes old and unused content from the CCMCache folder based on the specified number of days. It helps manage the storage space used by the CCMCache.

.PARAMETER Detect
Detects old and unused content in the CCMCache folder without deleting.

.PARAMETER Clean
Cleans old and unused content from the CCMCache folder. If no option is provided, this is the default.

.PARAMETER Help
Shows the help message.

.PARAMETER Days
Specifies the number of days to keep files in the CCMCache folder. Default is 30 days.

.EXAMPLE
.\Clear-CCMCache.ps1 -Detect

.EXAMPLE
.\Clear-CCMCache.ps1 -Clean

.EXAMPLE
.\Clear-CCMCache.ps1 -Help

.EXAMPLE
.\Clear-CCMCache.ps1 -Clean -Days 14

.EXAMPLE
.\Clear-CCMCache.ps1 -Detect -Days 14

.NOTES
This script is designed for administrators and developers who need to manage the CCMCache folder storage space.

#>

[CmdletBinding(PositionalBinding=$False)]
param(
    [switch]$Detect,
    [switch]$Clean,
    [switch]$Help,
    [int]$Days = 14
)

if ($PSVersionTable.PSVersion.Major -eq 7) {
    Write-Error "This script does not support PowerShell 7.x. Exiting..."
    exit
}
function Show-Help {
    @"
Usage: .\Clear-CCMCache.ps1 [-Detect] [-Clean] [-Help] [-Days ##]

-Detect : Detect old and unused content in the CCMCache folder.
-Clean  : Clean old and unused content from the CCMCache folder. If no option is provided, this is the default.
-Help   : Show this help message.
-Days   : Set the number of days to keep files in the CCMCache folder. Default is 30 days.

Examples:
  .\Clear-CCMCache.ps1 -Detect
  .\Clear-CCMCache.ps1 -Clean
  .\Clear-CCMCache.ps1 -Help
  .\Clear-CCMCache.ps1 -Clean -Days 14
  .\Clear-CCMCache.ps1 -Detect -Days 14
"@
}

if ($Help) {
    Show-Help
    exit
}

if (-not $Detect -and -not $Clean) {
    $Clean = $true
}
 
# Get CCMCache path
$CachePath = ([wmi]"ROOT\ccm\SoftMgmtAgent:CacheConfig.ConfigKey='Cache'").Location

# Test if the $CachePath was read successfuly retrieved and exit script if not
if ($null -ne $CachePath -and $CachePath -ne '')
{
 # Continue with the script execution

# Get items not referenced for more than the specified days
$OldCache = Get-WmiObject -Query "SELECT * FROM CacheInfoEx" -Namespace "ROOT\ccm\SoftMgmtAgent" | Where-Object { ([datetime]::Now - ([System.Management.ManagementDateTimeConverter]::ToDateTime($_.LastReferenced))).Days -gt $Days }

if ($Detect) {
    # Report old items
    if ($OldCache) { $false } else { $true }
}

if ($Clean) {
    # Delete items on disk
    echo $OldCache
    $OldCache | ForEach-Object { Remove-Item -Path $_.Location -Recurse -Force -ErrorAction SilentlyContinue }

    # Delete items in WMI
    $OldCache | Remove-WmiObject

    # Get all cached items from disk
    write-host $CacheFoldersDisk
    $CacheFoldersDisk = (Get-ChildItem $CachePath).FullName

    # Get all cached items from WMI
    $CacheFoldersWMI = Get-WmiObject -Query "SELECT * FROM CacheInfoEx" -Namespace "ROOT\ccm\SoftMgmtAgent"

    # Remove orphaned folders from disk
    write-host $CacheFoldersDisk
    $CacheFoldersDisk | ForEach-Object { if ($_ -notin $CacheFoldersWMI.Location) { Remove-Item -Path $_ -Recurse -Force -ErrorAction SilentlyContinue } }

    # Remove orphaned WMI objects
    $CacheFoldersWMI | ForEach-Object { if ($_.Location -notin $CacheFoldersDisk) { $_ | Remove-WmiObject } }
}
} else { #exit if $Cachepath not retrieved
    Write-Error "Failed to set CachePath. Exiting script."
    exit 1
}


```

**Comment by virtualox** [comment]
Thank you for identifying the critical issue and providing a comprehensive solution. Your enhancements to the script significantly improve its safety and reliability. I have reviewed and incorporated your changes into the main branch. Additionally, I added a `#Requires -RunAsAdministrator` directive to ensure the script always runs with the necessary privileges.

Your contribution is greatly appreciated!
